pyquil/quilbase.py

```diff
@@ -64,26 +70,42 @@
 
 from dataclasses import dataclass
 
+import qcs_sdk.quil.instructions as quil_rs
 
-class AbstractInstruction(object):
+
+class _InstructionMeta(abc.ABCMeta):
```

Comment:
This is a very hacky workaround for [quil-rs](https://github.com/rigetti/quail-rs/issues/148). 

I'll call out (at least) two things that are probably not great about this:

1. Performance:

`quil_rs.Instruction.__new__` works by attempting to initialize one of every instruction variant until it is successful, or returns an error if it failed for all of them. I haven't benchmarked this, and I don't think it's likely to make a difference in practice, but worth noting.

2. `quil_rs` instructions don't have `out()`

The `out()` method was part of the `AbstractInstruction` class, so anything passing `isinstance(x, AbstractInstruction)` would be recognized as having it.

That said, `AbstractInstruction` didn't provide any real implementation of `out()` and though it probably was intended to be used as an abstract base class, it isn't one, so `out()` wasn't ever a real guarantee. Since that was the case, I removed it from `AbstractInstruction`.

This won't break any user scripts as far as I can tell. If they were relying on an `out()` method, it must be from an instruction that actually implemented it. Tools like `mypy` or `pyright` might complain if one uses `out()` when the only thing known about the type is that it is an `AbstractInstruction`, but that isn't a hard break (and an easy one to fix).

Reply:
(1) seems, like you say, "not great" and I imagine we may have to circle around and address this. That's just a feeling and I don't have any better suggestions.

(2) I agree with your analysis - this should be just fine as you have it, for the reason you name: any `out()` call someone was already making will still be supported.